src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -324,42 +324,78 @@ impl KvmVcpu {
-
-        for msr_index_chunk in msr_index_list.chunks(KVM_MAX_MSR_ENTRIES) {
-            let mut msrs = Msrs::new(msr_index_chunk.len())?;
-            let msr_entries = msrs.as_mut_slice();
-            assert_eq!(msr_index_chunk.len(), msr_entries.len());
-            for (pos, index) in msr_index_chunk.iter().enumerate() {
-                msr_entries[pos].index = *index;
-            }
-
-            let expected_nmsrs = msrs.as_slice().len();
-            let nmsrs = self
-                .fd
-                .get_msrs(&mut msrs)
-                .map_err(KvmVcpuError::VcpuGetMsrs)?;
-            if nmsrs != expected_nmsrs {
-                return Err(KvmVcpuError::VcpuGetMsr(msr_index_chunk[nmsrs]));
-            }
-
-            msr_chunks.push(msrs);
+    fn get_msr_chunks(
+        &self,
+        mut msr_index_iter: impl ExactSizeIterator<Item = u32>,
+    ) -> Result<Vec<Msrs>, KvmVcpuError> {
+        let num_chunks = msr_index_iter.len() / KVM_MAX_MSR_ENTRIES;
+        let excess = msr_index_iter.len() % KVM_MAX_MSR_ENTRIES;
+
+        // Allocate 1 more element in case excess is not 0
+        let mut msr_chunks: Vec<Msrs> = Vec::with_capacity(num_chunks + 1);
+
+        for _ in 0..num_chunks {
+            let chunk = self.get_msr_chunk(&mut msr_index_iter, KVM_MAX_MSR_ENTRIES)?;
+            msr_chunks.push(chunk);
+        }
+        if excess != 0 {
+            let chunk = self.get_msr_chunk(&mut msr_index_iter, excess)?;
+            msr_chunks.push(chunk);
         }
 
         Self::fix_zero_tsc_deadline_msr(&mut msr_chunks);
 
         Ok(msr_chunks)
     }
 
+    /// Get single MSR chunk for the given MSR index iterator with
+    /// specified length. Iterator should have enough elements
+    /// to fill the chunk with indices, otherwise KVM will
+    /// return an error when processing half filled chunk.
+    ///
+    /// # Arguments
+    ///
+    /// * `msr_index_iter`: Iterator over MSR indices.
+    /// * `chunk_size`: Lenght of a chunk.
+    ///
+    /// # Errors
+    ///
+    /// * When [`kvm_bindings::Msrs::new`] returns errors.
+    /// * When [`kvm_ioctls::VcpuFd::get_msrs`] returns errors.
+    /// * When the return value of [`kvm_ioctls::VcpuFd::get_msrs`] (the number of entries that
+    ///   could be gotten) is less than expected.
+    pub fn get_msr_chunk(
+        &self,
+        msr_index_iter: impl Iterator<Item = u32>,
+        chunk_size: usize,
+    ) -> Result<Msrs, KvmVcpuError> {
+        let chunk_iter = msr_index_iter.take(chunk_size);
+
+        let mut msrs = Msrs::new(chunk_size)?;
+        let msr_entries = msrs.as_mut_slice();
+        for (pos, msr_index) in chunk_iter.enumerate() {
+            msr_entries[pos].index = msr_index;
+        }
+
+        let nmsrs = self
+            .fd
+            .get_msrs(&mut msrs)
+            .map_err(KvmVcpuError::VcpuGetMsrs)?;
+        if nmsrs != chunk_size {
+            Err(KvmVcpuError::VcpuGetMsr(msrs.as_slice()[nmsrs].index))
```

Comment:
Mh, I'm assuming `msrs.as_slice()[nmsrs]` is supposed to be the first MSR from the request that failed to be retrieved? Can we add a comment about that? I had to scratch my head a bit about why we're looking at `nmsrs`, because at first I thought we were trying to get the last MSR that _was_ successfully retrieved (in which case this would've needed to be `nmsrs - 1` and would break if `nmsrs == 0`).

Reply:
I had the same question at first, so I asked @zulinx86 about it. It seems the `get_msrs` does return a number of msrs set successfully. In this case the `nmsrs != chunk_size` check is correct. The `msrs.as_slice()[nmsrs]` is also correct because if some msrs could not be retrieved `nmsrs` will be an index of the failed `msr`. Example: you ask for 2 msrs, if KVM returns 1 because it could only retrieve mrs with index 0, then 1 points to the problematic msr. If KVM returns 2, than everything is fine.

Added a comment about this.

Reply:
Thank you (both for the comment and explanation)!